sys/shell/commands/sc_gnrc_rpl.c

```diff
@@ -206,31 +206,102 @@ int _gnrc_rpl_send_dis(void)
+     * the stats are grouped such that closely related values are read together.
+     * Hence, related metrics will always refer to the same state of the stats.
+     */
     puts(  "Statistics        (ucast) RX / TX                  RX / TX (mcast)");
-    printf("DIO     #packets: %10" PRIu32 " / %-10" PRIu32 "  %10" PRIu32 " / %-10" PRIu32 "\n",
-           gnrc_rpl_netstats.dio_rx_ucast_count, gnrc_rpl_netstats.dio_tx_ucast_count,
-           gnrc_rpl_netstats.dio_rx_mcast_count, gnrc_rpl_netstats.dio_tx_mcast_count);
-    printf("DIO       #bytes: %10" PRIu32 " / %-10" PRIu32 "  %10" PRIu32 " / %-10" PRIu32 "\n",
-           gnrc_rpl_netstats.dio_rx_ucast_bytes, gnrc_rpl_netstats.dio_tx_ucast_bytes,
-           gnrc_rpl_netstats.dio_rx_mcast_bytes, gnrc_rpl_netstats.dio_tx_mcast_bytes);
-    printf("DIS     #packets: %10" PRIu32 " / %-10" PRIu32 "  %10" PRIu32 " / %-10" PRIu32 "\n",
-           gnrc_rpl_netstats.dis_rx_ucast_count, gnrc_rpl_netstats.dis_tx_ucast_count,
-           gnrc_rpl_netstats.dis_rx_mcast_count, gnrc_rpl_netstats.dis_tx_mcast_count);
-    printf("DIS       #bytes: %10" PRIu32 " / %-10" PRIu32 "  %10" PRIu32 " / %-10" PRIu32 "\n",
-           gnrc_rpl_netstats.dis_rx_ucast_bytes, gnrc_rpl_netstats.dis_tx_ucast_bytes,
-           gnrc_rpl_netstats.dis_rx_mcast_bytes, gnrc_rpl_netstats.dis_tx_mcast_bytes);
-    printf("DAO     #packets: %10" PRIu32 " / %-10" PRIu32 "  %10" PRIu32 " / %-10" PRIu32 "\n",
-           gnrc_rpl_netstats.dao_rx_ucast_count, gnrc_rpl_netstats.dao_tx_ucast_count,
-           gnrc_rpl_netstats.dao_rx_mcast_count, gnrc_rpl_netstats.dao_tx_mcast_count);
-    printf("DAO       #bytes: %10" PRIu32 " / %-10" PRIu32 "  %10" PRIu32 " / %-10" PRIu32 "\n",
-           gnrc_rpl_netstats.dao_rx_ucast_bytes, gnrc_rpl_netstats.dao_tx_ucast_bytes,
-           gnrc_rpl_netstats.dao_rx_mcast_bytes, gnrc_rpl_netstats.dao_tx_mcast_bytes);
-    printf("DAO-ACK #packets: %10" PRIu32 " / %-10" PRIu32 "  %10" PRIu32 " / %-10" PRIu32 "\n",
-           gnrc_rpl_netstats.dao_ack_rx_ucast_count, gnrc_rpl_netstats.dao_ack_tx_ucast_count,
-           gnrc_rpl_netstats.dao_ack_rx_mcast_count, gnrc_rpl_netstats.dao_ack_tx_mcast_count);
-    printf("DAO-ACK   #bytes: %10" PRIu32 " / %-10" PRIu32 "  %10" PRIu32 " / %-10" PRIu32 "\n",
-           gnrc_rpl_netstats.dao_ack_rx_ucast_bytes, gnrc_rpl_netstats.dao_ack_tx_ucast_bytes,
-           gnrc_rpl_netstats.dao_ack_rx_mcast_bytes, gnrc_rpl_netstats.dao_ack_tx_mcast_bytes);
+    {
+        unsigned irq_state = irq_disable();
+        uint32_t rx_ucast = gnrc_rpl_netstats.dio_rx_ucast_count;
+        uint32_t tx_ucast = gnrc_rpl_netstats.dio_tx_ucast_count;
+        uint32_t rx_mcast = gnrc_rpl_netstats.dio_rx_mcast_count;
+        uint32_t tx_mcast = gnrc_rpl_netstats.dio_tx_mcast_count;
+        irq_restore(irq_state);
+        printf("DIO     #packets: %10" PRIu32 " / %-10" PRIu32 "  %10" PRIu32 " / %-10" PRIu32 "\n",
+               rx_ucast, tx_ucast, rx_mcast, tx_mcast);
```

Comment:
I rarely advocate for a macro, but given the amount of code-duplication (and the potential for copy-paste errors), how about introducing a compile-unit-local macro for this pattern?

```C
#define RPL_STAT_PRINT(netstats, pkt_name, pkt, stat_name, stat) { \
    unsigned irq_state = irq_disable(); \
    uint32_t rx_ucast = (netstats).pkt ## _rx_ucast_ ## stat; \
    uint32_t tx_ucast = (netstats).pkt ## _tx_ucast_ ## stat; \
    uint32_t rx_mcast = (netstats).pkt ## _rx_mcast_ ## stat; \
    uint32_t tx_mcast = (netstats).pkt ## _tx_mcast_ ## stat; \
    irq_restore(irq_state); \
    printf(type_name "     #" stat_name ": "
           "%10" PRIu32 " / %-10" PRIu32 "  %10" PRIu32 " / %-10" PRIu32 "\n", \
           rx_ucast, tx_ucast, rx_mcast, tx_mcast); \
}
```

If we are fine with changing the output, we can also just re-use `type` and `stat` in the output:

```C
    printf(#type "     #" #stat ": %10" PRIu32 " / %-10" PRIu32 "  %10" PRIu32 " / %-10" PRIu32 "\n", \
           rx_ucast, tx_ucast, rx_mcast, tx_mcast); \
```

Reply:
OK. But let's be quick before @kaspar030 notices ;)

Reply:
Wait, I can do this in C :)

Reply:
> Wait, I can do this in C :)

Even better!